openedx/core/djangoapps/certificates/api.py

```diff
@@ -19,6 +27,66 @@ def _enabled_and_instructor_paced(course):
     return False
 
 
+def certificates_viewable_for_course(course):
+    """
+    Returns True if certificates are viewable for any student enrolled in the course, False otherwise.
+    """
+    if course.self_paced:
+        return True
+    if (
+        course.certificates_display_behavior in ('early_with_info', 'early_no_info')
+        or course.certificates_show_before_end
+    ):
+        return True
+    if (
+        course.certificate_available_date
+        and course.certificate_available_date <= datetime.now(UTC)
+    ):
+        return True
+    if (
+        course.certificate_available_date is None
+        and course.has_ended()
+    ):
+        return True
+    return False
+
+
+def enrollment_is_verified(student, course_id):
+    """
+    Returns True if the student has a verified certificate enrollment in the course, False otherwise.
+    """
+    enrollment_mode, __ = CourseEnrollment.enrollment_mode_for_user(student, course_id)
+    return enrollment_mode in VERIFIED_CERTS_MODES
+
+
+def enrollment_is_active(student, course_id):
+    return CourseEnrollment.enrollment_mode_for_user(student, course_id)[1]
+
+
+def is_certificate_valid(certificate):
+    """
+    Returns True if the student has a valid, verified certificate for this course, False otherwise.
+    """
+    return enrollment_is_verified(certificate.user, certificate.course_id) and certificate.is_valid()
+
+
+def can_show_certificate_message(course, student, grade_summary, certificates_enabled_for_course):
+    if not (
+            (auto_certificate_generation_enabled() or certificates_enabled_for_course) and
+            enrollment_is_active(student, course.id) and
+            certificates_viewable_for_course(course) and
+            is_course_passed(course, grade_summary)
+    ):
+        return False
+    return True
+
+
+def is_course_passed(course, grade_summary):
+    nonzero_cutoffs = [cutoff for cutoff in course.grade_cutoffs.values() if cutoff > 0]
+    success_cutoff = min(nonzero_cutoffs) if nonzero_cutoffs else None
+    return success_cutoff and grade_summary['percent'] >= success_cutoff
```

Comment:
This is an abstraction violation with how passing grades is computed.  This code should be replaced with `course_grade.passed` (and not use the deprecated `grade_summary`).

Reply:
👍